### PR TITLE
Cleanup: broadcasting

### PR DESCRIPTION
**Pull request by anntzer**
- Rely more on broadcasting.
- Use power operator / `np.hypot` where appropriate.
- Use `np.pi` instead of `math.pi`.

### CODE REVIEW

File: examples/animation/unchained.py [hunk 1]
```diff
@@ -25,7 +25,7 @@
 # Generate random data
 data = np.random.uniform(0, 1, (64, 75))
 X = np.linspace(-1, 1, data.shape[-1])
-G = 1.5 * np.exp(-4 * X * X)
```
Comment (tacaswell) [inline examples/animation/unchained.py]: In most cases I think this construct is marginally faster.

Comment (anntzer) [inline examples/animation/unchained.py]:
```
# float is probably the most common case
In [8]: x = np.arange(100000).astype(float)

In [9]: %timeit x * x
10000 loops, best of 3: 65.5 µs per loop

In [10]: %timeit x ** 2
10000 loops, best of 3: 66.4 µs per loop
```
(and you can easily get them reversed on another run).

File: lib/matplotlib/axes/_axes.py [hunk 2]
```diff
@@ -1997,20 +1998,14 @@ def bar(self, left, height, width=0.8, bottom=None, **kwargs):
...
 
-        def make_iterable(x):
-            if not iterable(x):
-                return [x]
-            else:
-                return x
-
         # make them safe to take len() of
         _left = left
-        left = make_iterable(left)
```
Comment (tacaswell) [inline lib/matplotlib/axes/_axes.py]: Does this break unit handling?  It is probably better to leave the duck-ish typing here and not force to numpy

Comment (anntzer) [inline lib/matplotlib/axes/_axes.py]: I don't know what unit handling is actually supposed to do, but in any case if an object array (or list of objects, or single object) is passed in then atleast_1d will return an object array.

File: lib/matplotlib/projections/geo.py [hunk 3]
```diff
@@ -388,8 +388,8 @@ def transform_non_affine(self, xy):
 
             quarter_x = 0.25 * x
             half_y = 0.5 * y
-            z = np.sqrt(1.0 - quarter_x*quarter_x - half_y*half_y)
-            longitude = 2 * np.arctan((z*x) / (2.0 * (2.0*z*z - 1.0)))
+            z = np.sqrt(1 - quarter_x * quarter_x - half_y * half_y)
```
Comment (QuLogic) [inline lib/matplotlib/projections/geo.py]: No `**2` again?

File: lib/matplotlib/tests/test_axes.py [hunk 4]
```diff
@@ -1445,8 +1445,8 @@ def bump(a):
             y = 2 * np.random.random() - .5
             z = 10 / (.1 + np.random.random())
             for i in range(m):
-                w = (i / float(m) - y) * z
-                a[i] += x * np.exp(-w * w)
+                w = (i / m - y) * z
```
Comment (QuLogic) [inline lib/matplotlib/tests/test_axes.py]: Could be written as a NumPy expression instead of a loop.

File: lib/mpl_toolkits/mplot3d/axes3d.py [hunk 5]
```diff
@@ -2258,18 +2258,7 @@ def scatter(self, xs, ys, zs=0, zdir='z', s=20, c=None, depthshade=True,
...
-        if xs.size != ys.size:
-            raise ValueError("Arguments 'xs' and 'ys' must be of same size.")
-        if xs.size != zs.size:
-            if zs.size == 1:
-                zs = np.tile(zs[0], xs.size)
-            else:
-                raise ValueError(("Argument 'zs' must be of same size as 'xs' "
-                    "and 'ys' or of size 1."))
-
+        xs, ys, zs = np.broadcast_arrays(*map(np.ma.ravel, [xs, ys, zs]))
```
Comment (QuLogic) [inline lib/mpl_toolkits/mplot3d/axes3d.py]: This _possibly_ makes any error about incorrect shapes a bit more obscure.

File: examples/pylab_examples/scatter_masked.py [hunk 6]
```diff
@@ -7,17 +7,17 @@
...
 r0 = 0.6
-x = 0.9*np.random.rand(N)
-y = 0.9*np.random.rand(N)
-area = np.pi*(10 * np.random.rand(N))**2  # 0 to 10 point radiuses
+x = 0.9 * np.random.rand(N)
+y = 0.9 * np.random.rand(N)
+area = np.pi * (10 * np.random.rand(N)) ** 2  # 0 to 10 point radiuses
 c = np.sqrt(area)
-r = np.sqrt(x*x + y*y)
+r = np.sqrt(x * x + y * y)
```
Comment (anntzer) [inline examples/pylab_examples/scatter_masked.py]: I don't think we have to use hypot *every time*, especially in examples where the function may be somewhat unknown.

File: lib/matplotlib/backends/backend_wx.py [hunk 7]
```diff
@@ -318,7 +318,7 @@ def draw_text(self, gc, x, y, s, prop, angle, ismath=False, mtext=None):
         if angle == 0.0:
             gfx_ctx.DrawText(s, x, y)
         else:
-            rads = angle / 180.0 * math.pi
+            rads = math.radians(angle)
```
Comment (anntzer) [inline lib/matplotlib/backends/backend_wx.py]: `math` only has `radians` and `degrees`, not `deg2rad` and `rad2deg`.  I seems that both names are used in the codebase (even before my earlier PR).

File: examples/api/custom_projection_example.py [hunk 8]
```diff
@@ -440,8 +440,8 @@ def transform_non_affine(self, xy):
 
             quarter_x = 0.25 * x
             half_y = 0.5 * y
-            z = np.sqrt(1.0 - quarter_x*quarter_x - half_y*half_y)
-            longitude = 2 * np.arctan((z*x) / (2.0 * (2.0*z*z - 1.0)))
+            z = np.sqrt(1 - quarter_x * quarter_x - half_y * half_y)
```
Comment (anntzer) [inline examples/api/custom_projection_example.py]: Sure.

File: lib/matplotlib/tests/test_axes.py (hunk 4 above)
Comment (anntzer) [inline lib/matplotlib/tests/test_axes.py]: Sure.

File: examples/animation/unchained.py (hunk 1 above)
Comment (tacaswell) [inline examples/animation/unchained.py]: interesting.  That is an apparently wrong rule of thumb I have been using for a while.

Comment (anntzer) [inline examples/animation/unchained.py]: Why would you think this would be the case?  (I am honestly puzzled.)

File: lib/mpl_toolkits/mplot3d/axes3d.py (hunk 5 above)
Comment (anntzer) [inline lib/mpl_toolkits/mplot3d/axes3d.py]: `broadcast_to` explicitly prints the nonmatching shape in the error message, but not `broadcast_arrays`.  I'll open an issue on numpy but I don't think this should hold up this PR.

(Note how just below we gain checking on the size of dx/dy/dz, which was not present before.)

File: lib/matplotlib/backends/backend_wx.py (hunk 7 above)
Comment (QuLogic) [inline lib/matplotlib/backends/backend_wx.py]: Oops, I meant to go back and check whether that was `math` or `np` and forgot about it.

File: lib/matplotlib/axes/_axes.py (hunk 2 above)
Comment (anntzer) [inline lib/matplotlib/axes/_axes.py]: This seems due to a useless and incorrect implementation of `__array__`.  Tried to fix the example...

**Comment by efiring** [comment]
Looks like lots of nice cleanups here, once again.  In future, however, I suggest that you reconsider your policy of always surrounding operators by spaces.  This is not required by PEP 8, and although I am a fan of white space, I think that in many cases omitting spaces around operators improves readability.  E.g., `a * b**2` is better than `a * b ** 2` and `a*b + c*d` is better than `a * b + c * d`.  This is mainly for variables with very short names.

File: examples/animation/unchained.py (hunk 1 above)
Comment (efiring) [inline examples/animation/unchained.py]: It dates back to the early days of computers and compilers, when computers were slow and compilers were not so bright.  Multiplication is faster than taking a power, in general.  But now compilers recognize things like this and find the best algorithm.  In the case of numpy, I'm pretty sure there is explicit internal optimization of small integer powers so that we wouldn't have to use that ancient manual optimization.

**Comment by anntzer** [comment]
Sounds like a reasonable policy, thanks for the notice.

**Comment by anntzer** [comment]
Looks like there the build failure comes from the newly released sphinx 1.5.

**Comment by QuLogic** [comment]
Yep, #7569.

**Comment by anntzer** [comment]
I got rid of all the trivial whitespace changes.  (Lines with both whitespace changes and nonwhitespace changes kept their changes.)

**Comment by ianthomas23** [comment]
@anntzer Thankyou, but you do need to be consistent.  Changing some is worse than changing all or none.

**Comment by anntzer** [comment]
I think it's all fairly consistent now.

**Comment by anntzer** [comment]
There seems to be some issues left with the unit handling code that I need to look into.

File: examples/animation/unchained.py (hunk 1 above)
Comment (tacaswell) [inline examples/animation/unchained.py]: I respectfully request the children remain off of my lawn. :smiling_imp:

File: examples/pylab_examples/annotation_demo.py [hunk 9]
```diff
@@ -91,7 +91,7 @@
 fig = plt.figure()
 ax = fig.add_subplot(111, projection='polar')
 r = np.arange(0, 1, 0.001)
-theta = 2*2*np.pi*r
+theta = 4 * np.pi * r
```
Comment (tacaswell) [inline examples/pylab_examples/annotation_demo.py]: The `2*2` here may have been pedagogical (ex `2 * tau` )

File: lib/matplotlib/stackplot.py [hunk 10]
```diff
@@ -55,10 +55,7 @@ def stackplot(axes, x, *args, **kwargs):
     element in the stacked area plot.
     """
 
-    if len(args) == 1:
-        y = np.atleast_2d(*args)
-    elif len(args) > 1:
-        y = np.row_stack(args)
+    y = np.row_stack(args)
```
Comment (tacaswell) [inline lib/matplotlib/stackplot.py]: These are not equivalent

```python
In [18]: np.atleast_2d([1, 2, 3])
Out[18]: array([[1, 2, 3]])

In [19]: np.row_stack([1, 2, 3])
Out[19]: 
array([[1],
       [2],
       [3]])


```

File: lib/matplotlib/tests/test_axes.py [hunk 11]
```diff
@@ -1439,9 +1439,7 @@ def bump(a):
             x = 1 / (.1 + np.random.random())
             y = 2 * np.random.random() - .5
             z = 10 / (.1 + np.random.random())
-            for i in range(m):
-                w = (i / float(m) - y) * z
-                a[i] += x * np.exp(-w * w)
+            a += x * np.exp(-((np.arange(m) / m - y) * z) ** 2)
         a = np.zeros((m, n))
         for i in range(n):
             for j in range(5):
```
Comment (tacaswell) [inline lib/matplotlib/tests/test_axes.py]: Why is this inner loop here?

File: examples/pylab_examples/annotation_demo.py (hunk 9 above)
Comment (anntzer) [inline examples/pylab_examples/annotation_demo.py]: Reverted; however I chose to leave the use of `**2` instead of `X * X` above (unless you feel strongly about it).

File: lib/matplotlib/tests/test_axes.py (hunk 11 above)
Comment (anntzer) [inline lib/matplotlib/tests/test_axes.py]: Because whoever came up with that example decided to add some random numbers five times to each column rather than only once...

File: lib/matplotlib/stackplot.py (hunk 10 above)
Comment (anntzer) [inline lib/matplotlib/stackplot.py]: No, it's `np.atleast_2d(*args)` (note the unpack) and *only* in the case where `len(args) == 1`.

Say `args` has shape (after casting to an array) `(1, x, y, ...)`.
- `np.atleast_2d(*args) == np.atleast_2d(<shape (x, y, ...)>) == <shape (x, y, ...)>`
- `np.row_stack(args)` also has shape (x, y, ...).

In the case `args` is 2D (shape (1, x)), both expressions result in a shape (1, x) as well.

File: lib/matplotlib/axes/_axes.py (hunk 2 above)
Comment (anntzer) [inline lib/matplotlib/axes/_axes.py]: `atleast_1d` calls `asanyarray` so it passes subclasses through with no problem.  It is true that pandas Series do not subclass numpy arrays and get copied, but given that we're going to generate one Rectangle object per entry in the Series I'd say the additional cost of the copy is rather small.

Comment (anntzer) [inline lib/matplotlib/axes/_axes.py]: Also, note that the current implementation is actually buggy: `bar([1, 2], [3, 4], width=[.8])` works but `bar([1, 2], [3, 4], width=np.array([.8]))` currently raises an exception due to the fact that `width *= nbars` works elementwise here.  (Can you open a separate issue if you don't think this PR can get merged any time soon, so that it doesn't get lost?)

Comment (anntzer) [inline lib/matplotlib/axes/_axes.py]: Edit: Actually asarray doesn't even copy the underlying buffer for pandas series:
```
In [1]: s = pd.Series([1, 2, 3]); t = np.asarray(s); t[0] = 42; s
Out[1]: 
0    42
1     2
2     3
dtype: int64
```
so the point is moot.

File: lib/matplotlib/stackplot.py (hunk 10 above)
Comment (tacaswell) [inline lib/matplotlib/stackplot.py]: 🐑

File: lib/matplotlib/tests/test_axes.py (hunk 11 above)
Comment (tacaswell) [inline lib/matplotlib/tests/test_axes.py]: 🐑 Yeah, I am greatly confused by the local operating in-place function...

File: examples/pylab_examples/annotation_demo.py (hunk 9 above)
Comment (tacaswell) [inline examples/pylab_examples/annotation_demo.py]: no, @efiring explained that one correctly.  If it isn't faster, use the more obvious way to write it.

File: lib/matplotlib/tests/test_axes.py (hunk 11 above)
Comment (anntzer) [inline lib/matplotlib/tests/test_axes.py]: `stackplot_demo2` (from which this function comes) has the slightly more helpful docstring "Return *n* random Gaussian mixtures, each of length *m*."  I can copy it there, or just inline the function (also in the demo), let me know if you have a preference.

**Comment by anntzer** [comment]
I included a fix to #6820, given that I was already halfway there...
The test was copied from #7644, which specifically addresses that issue from scratch.

**Comment by anntzer** [comment]
Note to self: I need to backport a recent version of broadcast_arrays too, in order to have subok support (for passing masked arrays through).

Not ready for merging anymore...

**Comment by anntzer** [comment]
Eliminated the use of subok, and fixed a previously existing bug in the handling of masked arrays by fill_between (which was kept through invalid test images...).  See the comment of the 4th commit ("Broadcast in fill_between{,x} as well.")

Ready for review again (hopefully).

**Comment by QuLogic** [comment]
Ach, so many conflicts...

**Comment by anntzer** [comment]
should be fixed now...

File: lib/matplotlib/colors.py [hunk 12]
```diff
@@ -793,25 +795,23 @@ def __init__(self, colors, name='from_list', N=None):
...
         self.monochrome = False  # True only if all colors in map are
                                  # identical; needed for contouring.
         if N is None:
-            N = len(self.colors)
+            self.colors = colors
+            N = len(colors)
         else:
-            if isinstance(self.colors, six.string_types):
-                self.colors = [self.colors] * N
+            if isinstance(colors, six.string_types):
```
Comment (QuLogic) [inline lib/matplotlib/colors.py]: This nested `if` could be flattened, I think.

File: lib/mpl_toolkits/mplot3d/axes3d.py [hunk 13]
```diff
@@ -11,25 +11,23 @@
...
 
+import math
 import warnings
 
 import numpy as np
-import matplotlib.axes as maxes
+
+from matplotlib import (
+    axes as maxes, cbook, collections as mcoll, colors as mcolors, docstring,
+    scale as mscale, transforms as mtransforms)
```
Comment (QuLogic) [inline lib/mpl_toolkits/mplot3d/axes3d.py]: Don't really like this multiple-import-with-renaming thing.

File: lib/mpl_toolkits/mplot3d/axes3d.py [hunk 14]
```diff
@@ -2476,43 +2459,17 @@ def bar3d(self, x, y, z, dx, dy, dz, color=None,
...
+            np.atleast_1d(x), y, z, dx, dy, dz)
+        minx = np.min(x)
+        maxx = np.max(x + dx)
+        miny = np.min(y)
+        maxy = np.max(y + dy)
+        minz = np.min(z)
+        maxz = np.max(z + dz)
 
         polys = []
         for xi, yi, zi, dxi, dyi, dzi in zip(x, y, z, dx, dy, dz):
```
Comment (QuLogic) [inline lib/mpl_toolkits/mplot3d/axes3d.py]: I have a feeling this loop could be written as array operations as well.

**Comment by anntzer** [comment]
what's the "sword" smiley? :)

File: examples/pylab_examples/pythonic_matplotlib.py [hunk 15]
```diff
@@ -75,7 +75,7 @@
 
 
 ax2 = fig.add_subplot(212)
-ax2.plot(t, sin(2*2*pi*t))
+ax2.plot(t, sin(4 * pi * t))
```
Comment (anntzer) [inline examples/pylab_examples/pythonic_matplotlib.py]: sure

File: lib/mpl_toolkits/mplot3d/axes3d.py (hunk 13 above)
Comment (anntzer) [inline lib/mpl_toolkits/mplot3d/axes3d.py]: sure

File: lib/mpl_toolkits/mplot3d/axes3d.py (hunk 14 above)
Comment (anntzer) [inline lib/mpl_toolkits/mplot3d/axes3d.py]: it can almost certainly be vectorized, but you get to figure out how to do it :-)

File: lib/matplotlib/colors.py (hunk 12 above)
Comment (anntzer) [inline lib/matplotlib/colors.py]: yes, but i don't think this would make things more legible (N is None and N is not None are two logical alternatives)

File: lib/matplotlib/axes/_axes.py [hunk 16]
```diff
@@ -4879,22 +4816,11 @@ def fill_betweenx(self, y, x1, x2=0, where=None,
...
-
         if where is None:
-            where = np.ones(len(y), np.bool)
-        else:
-            where = np.asarray(where, np.bool)
-
-        if not (y.shape == x1.shape == x2.shape == where.shape):
-            raise ValueError("Argument dimensions are incompatible")
+            where = True
+        where = where & ~reduce(np.logical_or, map(np.ma.getmask, [y, x1, x2]))
```
Comment (anntzer) [inline lib/matplotlib/axes/_axes.py]: but then I can't use map.  not convinced it would be more legible.

File: examples/units/basic_units.py [hunk 17]
```diff
@@ -4,6 +4,8 @@
 ===========
 
 """
+import six
```
Comment (anntzer) [inline examples/units/basic_units.py]: IIRC there was some issue with how numpy broadcasting would interact with "poorly" designed unit classes.

File: lib/matplotlib/axes/_axes.py (hunk 16 above)
Comment (QuLogic) [inline lib/matplotlib/axes/_axes.py]: Thinking more about `reduce` "going away" in Python 3 more than legibility, but I guess we can just do `from functools import reduce` instead of `six.moves`.

File: lib/mpl_toolkits/mplot3d/axes3d.py (hunk 14 above)
Comment (QuLogic) [inline lib/mpl_toolkits/mplot3d/axes3d.py]: Oh, it's `extend`, not `append` like I first thought, which is a bit more complicated.

**Comment by QuLogic** [comment]
@tacaswell uses ⚔️ for conflicts...

File: lib/matplotlib/axes/_axes.py (hunk 16 above)
Comment (anntzer) [inline lib/matplotlib/axes/_axes.py]: changed to functools.reduce

**Comment by anntzer** [comment]
Rebased.

**Comment by QuLogic** [comment]
Still LGTM, though maybe squash the import fixup into the relevant commits. Might also want to rebase just to get CircleCI+AppVeyor fixes too..

**Comment by anntzer** [comment]
I squashed-rebased everything as splitting the fixup commit across the rebase seems a bit tricky, plus everything is still more or less related.

**Comment by QuLogic** [comment]
I forgot; why did the test image change?

**Comment by anntzer** [comment]
Because the previous behavior was incorrect: see the leftmost green "triangle", which is did not go all the way to the left but now does.

**Comment by QuLogic** [comment]
Ah, I see it now.

**Comment by QuLogic** [comment]
Needs a rebase; I'm also going to dismiss @tacaswell's review which seems to be outdated.

**Comment by anntzer** [comment]
done

**Comment by dopplershift** [comment]
And of course in that time a PR went in that causes conflicts...

**Comment by anntzer** [comment]
and fixed again...

File: examples/lines_bars_and_markers/nan_test.py [hunk 18]
```diff
@@ -9,7 +9,7 @@
 import matplotlib.pyplot as plt
 
 t = np.arange(0.0, 1.0 + 0.01, 0.01)
-s = np.cos(2 * 2 * np.pi * t)
+s = np.cos(2 * 2*np.pi * t)
```
Comment (dopplershift) [inline examples/lines_bars_and_markers/nan_test.py]: Just curious: why is *this* the one where you don't have spaces around it?

Comment (dopplershift) [inline examples/lines_bars_and_markers/nan_test.py]: nm...I guess I see it now.

File: lib/matplotlib/projections/polar.py [hunk 19]
```diff
@@ -1,7 +1,10 @@
 from __future__ import (absolute_import, division, print_function,
                         unicode_literals)
 
-import collections
+import six
+
+from collections import OrderedDict
+import warnings
```
Comment (dopplershift) [inline lib/matplotlib/projections/polar.py]: Why is this import necessary?

Comment (dopplershift) [inline lib/matplotlib/projections/polar.py]: Same thing actually for six.

Comment (anntzer) [inline lib/matplotlib/projections/polar.py]: Probably some rebase made the thing irrelevant.  Fixed.

File: examples/animation/unchained.py (hunk 1 above)
Comment (eric-wieser) [reply examples/animation/unchained.py]: This is not a useful measurement, because numpy hard-codes `power(x, 2)` to fall back on `square(x)`.
